### PR TITLE
Adding a task scheduler for background tasks

### PR DESCRIPTION
**Pull request by Andreea-Lupu**
**What type of PR is this?**

**Which issue does this PR fix**:

**What does this PR do / Why do we need it**:

**If an issue # is not available please add repro steps and logs showing the issue**:

**Testing done on this change**:

**Automation added to e2e**:

**Will this break upgrades or downgrades?**

**Does this PR introduce any user-facing change?**:

```release-note

```

By submitting this pull request, I confirm that my contribution is made under the terms of the Apache 2.0 license.

### CODE REVIEW

**Comment by rchincha** [comment]
The general design is that any extension should be able to create a Task (which should be an interface) to the TaskScheduler.
The TaskScheduler shouldn't know what the task is, just that it handles it in the background and returns a status or calls a callback.

File: pkg/extensions/scheduler/task-scheduler.go [hunk 1]
```diff
@@ -0,0 +1,94 @@
...
+				return
+			case task := <-scheduler.tasksQ:
+				scheduler.wg.Add(1)
+				status, callback := task.DoWork(scheduler.wg)
+
+				if callback != nil {
+					_ = callback(status)
+				}
+
+				if status == Running {
```
Comment (andaaron) [inline pkg/extensions/scheduler/task-scheduler.go:65]: Why is this needed? Under what condition is the status still Running at this line? Could it be any other value?

When is a task moved from running to a different status? I think that logic should be encapsulated in this package.

File: pkg/extensions/scheduler/task-scheduler.go [hunk 2]
```diff
@@ -0,0 +1,94 @@
...
+	go func() {
+		for {
+			select {
+			case <-ctx.Done():
+				scheduler.StopScheduler()
+
+				return
+			case task := <-scheduler.tasksQ:
+				scheduler.wg.Add(1)
+				status, callback := task.DoWork(scheduler.wg)
```
Comment (andaaron) [inline pkg/extensions/scheduler/task-scheduler.go:59]: So the purpose of the scheduler is to execute tasks one by one?

File: pkg/extensions/extension_scrub.go [hunk 3]
```diff
@@ -4,30 +4,107 @@
...
+				for {
+					if errIterator != nil && !errors.Is(errIterator, io.EOF) {
+						break
+					}
+
+					if repo == "" {
+						break
+					}
+
+					scrubTask := scrub.NewScrubTask(periodicScrubT.imgStore, repo, periodicScrubT.log)
```
Comment (andaaron) [inline pkg/extensions/extension_scrub.go:95]: Can we have the task DoWork in a generic method, and the scrub logic be passed as a parameter of type function to it? Most of this method looks like code which would be duplicated in all DoWork implementations.

We would pass the actual "work"/scrub function as well as the callback function(s ?) for success or failure to the generic DoWork function. These work/context functions would be called using the same context as parameter (which would be instantiated just before calling DoWork for this specific task and passed on as a parameter), so if one is cancelled, all are cancelled.

File: pkg/extensions/extension_scrub.go [hunk 4]
```diff
@@ -4,30 +4,107 @@
...
+	return periodicScrub
+}
+
+func (periodicScrubT *periodicScrubTask) DoWork(wgScheduler *goSync.WaitGroup) (scheduler.Status, scheduler.Cb) {
+	ticker := time.NewTicker(periodicScrubT.interval)
+
+	go func() {
+		for {
+			select {
+			case <-periodicScrubT.cancel:
```
Comment (andaaron) [inline pkg/extensions/extension_scrub.go:75]: Why not use a context - also for handling timeouts and cancellation - see some examples here https://www.sohamkamani.com/golang/context-cancellation-and-values/

File: pkg/extensions/scrub/scrub.go [hunk 5]
```diff
@@ -39,5 +44,47 @@ func RunScrubRepo(imgStore storage.ImageStore, repo string, log log.Logger) {
...
 	}
 
-	log.Info().Msg(fmt.Sprintf("scrub completed for %s", path.Join(imgStore.RootDir(), repo)))
+	log.Info().Msg(fmt.Sprintf("scrub successfully completed for %s", path.Join(imgStore.RootDir(), repo)))
+
+	return nil
+}
+
+type Task struct {
+	*scheduler.AbstractTask
```
Comment (andaaron) [inline pkg/extensions/scrub/scrub.go:53]: What do we need the abstract task for?

File: pkg/extensions/scheduler/task-scheduler.go [hunk 6]
```diff
@@ -0,0 +1,88 @@
+package scheduler
```
[the remark this answers is not in the document]
Comment (andaaron) [reply pkg/extensions/scheduler/task-scheduler.go:1]: This will be under pkg/scheduler.

File: pkg/scheduler/scheduler.go [hunk 7]
```diff
@@ -0,0 +1,244 @@
...
+		interval:      interval,
+		done:          false,
+		priority:      priority,
+		lastRepo:      "",
+		taskGenerator: taskGenerator,
+	}
+
+	scheduler.generatorsLock.Lock()
+	heap.Push(&scheduler.generators, newGenerator)
+	scheduler.generatorsLock.Unlock()
```
Comment (peusebiu) [inline pkg/scheduler/scheduler.go:243]: use defer :D

File: pkg/scheduler/scheduler.go [hunk 8]
```diff
@@ -0,0 +1,244 @@
...
+				gen.done = false
+			} else {
+				return
+			}
+		} else {
+			return
+		}
+	}
+
+	task, repo, err := gen.taskGenerator.GenerateTask(gen.lastRepo)
```
Comment (peusebiu) [inline pkg/scheduler/scheduler.go:217]: should we call this item instead of repo? it can be anything right?

File: pkg/extensions/extension_scrub.go [hunk 9]
```diff
@@ -9,25 +9,59 @@ import (
...
+	imgStore storage.ImageStore
+	log      log.Logger
+}
+
+func (gen taskGenerator) GenerateTask(lastRepo string) (scheduler.Task, string, error) {
+	var repo string
+
+	var err error
+
+	if lastRepo == "" {
```
Comment (peusebiu) [inline pkg/extensions/extension_scrub.go:60]: lastRepo can be a taskGenerator attribute?

**Review by andaaron — COMMENTED** [review]
Really nice work!
I left some comments on the doc.

File: pkg/storage/local.go [hunk 10]
```diff
@@ -360,6 +361,106 @@ func (is *ImageStoreLocal) GetRepositories() ([]string, error) {
...
+	_, err := os.ReadDir(dir)
+	if err != nil {
+		is.log.Error().Err(err).Msg("failure walking storage root-dir")
+
+		return "", err
+	}
+
+	found := false
+	store := ""
+	err = filepath.Walk(dir, func(path string, info os.FileInfo, err error) error {
```
[the remark this answers is not in the document]
Comment (andaaron) [reply pkg/storage/local.go]: Let's stick to this for now. I'd rather we have this feature working instead of debating these details.

**Review by andaaron — APPROVED** [review]

**Review by rchincha — APPROVED** [review]
lgtm
